text/0122-sdk-hub-scope-merge.md

```diff
@@ -0,0 +1,475 @@
+
+The [Global Scope](#global-scope) is truly global and can be kept in a global place.
+
+### Global Scope
+
+In addition to the current scope, there will also be a new special scope, the **Global Scope**.
+
+When you call `Sentry.init()`, you will immediately have a current scope - an initial scope.
+In the current/previous system, if you want to set global data (e.g. tags that should apply to _all_ events),
+you have to do this right after calling `Sentry.init()`, on the initial scope, so this data will be correctly inherited by all downstream scopes:
+
+```js
+Sentry.init();
+
+// Scope is the initial scope - the scope right after calling init()
+cont scope = Sentry.getCurrentHub().getCurrentScope();
+scope.setTag('revision', 'abc');
+```
+
+This becomes especially tricky if you want to set global scope data based on async processes, or in different files.
+The global scope should fix this by making it very clear how to setup data/processors that should apply to _all_ events, no matter where you define them:
+
+```js
+Sentry.init();
+setupGlobalData();
+
+// Somewhere else
+async function setupGlobalData() {
+  const revision = await getRevision();
+  // No need to worry or care about where this is called, this will _always_ work!
+  Sentry.getGlobalScope().setTag("revision", "abc");
+}
+```
+
+The global scope is _not_ the initial scope, but a separate, special scope. The global scope is the same for all clients - if you care about per-client isolation, you should not use the global scope.
+
+You can get the current global scope via `getGlobalScope()`. There _may_ be a function `setGlobalScope(scope)` to replace the global scope - or SDKs can decide that there is no need to replace the global scope, you can only mutate it. 
+
+If you call `getGlobalScope()` before a client is initialized, we should still get a global scope back. This way, we can avoid race conditions if mutating the global scope before `Sentry.init()` is called.
+
+The reason that the global scope is not the same as the initial scope of a client, is that you cannot accidentally mutate it - nothing ever inherits off the global scope.
+
+### How to document global scopes
+
+Global scopes will be documented in a new section under "Enriching Events > Global Data", which will explain how to use `getGlobalScope()` and when/how data added there will be applied.
+
+### Isolation Scope
+
+Furthermore, there are also **Isolation Scopes**.
+
+An isolation scope is similar to a global scope, but tied to a specific section of your app/code that should be isolated.
+In a server side language, an isolation scope will commonly be created per request. Each request gets an isolated scope that you can assign data to, which should be applied to all events that are processed in this request. However, an isolation scope is not limited to requests - you could also have an isolation scope for a scheduled job, or for a zone in your UI (in theory).
+
```

Comment:
Maybe not necessary to implement IsolatedScopes in all SDKs. 

The dotnet SDK has an asynclocal scope manager that achieves this already. I'm not sure how unique that is to .NET though and whether equivalents exist in other languages/runtimes.

Reply:
Does that mean in dotnet you can have both an isolation scope (via asynclocal scope manager) and a separate current scope?

Reply:
Possibly but it's not really a problem we'd need to solve in .NET. In ASP.NET Core requests get handled by middleware which is async/await all the way through and so we use an [asynclocal scope container](https://github.com/getsentry/sentry-dotnet/blob/cd166964b81b745cbbcda914fc1a601bd3e964dd/src/Sentry/Internal/ScopeStack/AsyncLocalScopeStackContainer.cs) (that has affinity for these requests). 

https://github.com/getsentry/rfcs/blob/c4b1fb4fd2915246760156285393340af093d591/text/0122-sdk-hub-scope-merge.md?plain=1#L304

... this isn't really necessary in .NET as we already get that for free from asynclocal.

Reply:
I was trying to say that parts of static API interact with isolation scope and other parts interact with current scope. I'm guessing you have to propagate both using this asynclocal scope container. Not sure about .NET but there'll probably be some `Sentry.withScope()` and/or `Sentry.pushScope/popScope` API that forks current scope.

Reply:
> Not sure about .NET but there'll probably be some `Sentry.withScope()` and/or `Sentry.pushScope/popScope` API that forks current scope

There is yes. There's the [SentrySdk.ConfigureScope](https://docs.sentry.io/platforms/dotnet/enriching-events/scopes/#configuring-the-scope) which clones the current scope and pops the clone on the stack, then removes the clone once it's done (ie. when it's disposed). 

I guess we could replace that with `SentrySdk.GetIsolationScope` or similar but I don't understand why we would. It doesn't seem to solve a problem that any of our users have and it doesn't get us any closer to using OTEL for spans (at least not for the .NET SDK - I realise it may be a prerequisite to this for the JavaScript SDKs).